folding-schemes/src/folding/hypernova/mod.rs

```diff
@@ -25,3 +55,534 @@ impl<F: PrimeField> Witness<F> {
+        i_bytes.copy_from_slice(&self.i.into_bigint().to_bytes_le()[..8]);
+        let i_usize: usize = usize::from_le_bytes(i_bytes);
+
+        let z_i1 = self
+            .F
+            .step_native(i_usize, self.z_i.clone(), external_inputs.clone())?;
+
+        let u_i1_x = U_i1.hash(
+            &self.poseidon_config,
+            self.i + C1::ScalarField::one(),
+            self.z_0.clone(),
+            z_i1.clone(),
+        )?;
+
+        // u_{i+1}.x[1] = H(cf_U_{i+1})
+        let cf_u_i1_x: C1::ScalarField;
+
+        if self.i == C1::ScalarField::zero() {
+            cf_u_i1_x = self.cf_U_i.hash_cyclefold(&self.poseidon_config)?;
+
+            augmented_f_circuit = AugmentedFCircuit::<C1, C2, GC2, FC> {
+                _c2: PhantomData,
+                _gc2: PhantomData,
+                poseidon_config: self.poseidon_config.clone(),
+                ccs: self.ccs.clone(),
+                i: Some(C1::ScalarField::zero()),
+                i_usize: Some(0),
+                z_0: Some(self.z_0.clone()),
+                z_i: Some(self.z_i.clone()),
+                external_inputs: Some(external_inputs.clone()),
+                u_i_C: Some(self.u_i.C),
+                U_i: Some(self.U_i.clone()),
+                U_i1_C: Some(U_i1.C),
+                F: self.F.clone(),
+                x: Some(u_i1_x),
+                nimfs_proof: Some(nimfs_proof),
+
+                // cyclefold values
+                cf_u_i_cmW: None,
+                cf_U_i: None,
+                cf_x: Some(cf_u_i1_x),
+                cf_cmT: None,
+            };
+        } else {
+            let rho_Fq = C2::ScalarField::from_bigint(BigInteger::from_bits_le(&rho_bits))
+                .ok_or(Error::OutOfBounds)?;
+            // CycleFold part:
+            // get the vector used as public inputs 'x' in the CycleFold circuit
+            // cyclefold circuit for cmW
+            let cf_u_i_x = [
+                vec![rho_Fq],
+                get_cm_coordinates(&self.U_i.C),
+                get_cm_coordinates(&self.u_i.C),
+                get_cm_coordinates(&U_i1.C),
+            ]
+            .concat();
+
+            let cf_circuit = CycleFoldCircuit::<C1, GC1> {
+                _gc: PhantomData,
+                r_bits: Some(rho_bits.clone()),
+                p1: Some(self.U_i.clone().C),
+                p2: Some(self.u_i.clone().C),
+                x: Some(cf_u_i_x.clone()),
+            };
+
+            let (_cf_w_i, cf_u_i, cf_W_i1, cf_U_i1, cf_cmT, _) =
+                fold_cyclefold_circuit::<C1, GC1, C2, GC2, FC, CS1, CS2>(
+                    &self.poseidon_config,
+                    self.cf_r1cs.clone(),
+                    self.cf_cs_params.clone(),
+                    self.cf_W_i.clone(), // CycleFold running instance witness
+                    self.cf_U_i.clone(), // CycleFold running instance
+                    cf_u_i_x,
+                    cf_circuit,
+                )?;
```

Comment:
Should this instead return an object `CycleFoldInstance` or similar? 

On this way, we have less verbosity here and down there constructing the `AugmentedFCircuit`

Reply:
we could convert that into an object, but anyways in the following line it needs to be split again into separated parameters, which would be less simple than currently (on each parameter used would need to pass `struct.param0` instead of `param0` directly), so I'm not sure on it

Reply:
That's fair. I was thinking on having this struct inside of `AugmentedFCircuit` also. Such that construct it from it's params is much simpler and shorter.

Reply:
but they are not only used in the AugmentedFCircuit

Reply:
In the other places you can simply have getters. https://github.com/jbaublitz/getset is a widely used tool to generate them without code pollution nor complexity.

Reply:
If you think that this is not worth it, feel free to mark as resolved otherwise :)

Reply:
hmm but having getters would be even worse in terms of code complexity than just accessing the struct params, no?

Reply:
(eg. `struct.param0` vs `struct.get_param0()`)

Reply:
well, you never use `get_` prefix in rust. And also the macro removes all complexity from defining the functions. 
I just try to look for a way to reduce the massive code blocks here. 

As said, if you aren't convinced, let's just close this. It's no issues.